charts/ccsm-helm/test/operate/configmap_test.go

```diff
@@ -0,0 +1,57 @@
+package operate
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/gruntwork-io/terratest/modules/helm"
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	"github.com/gruntwork-io/terratest/modules/random"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type configmapTemplateTest struct {
+	suite.Suite
+	chartPath string
+	release   string
+	namespace string
+	templates []string
+}
+
+func TestConfigmapTemplate(t *testing.T) {
+	t.Parallel()
+
+	chartPath, err := filepath.Abs("../../")
+	require.NoError(t, err)
+
+	suite.Run(t, &configmapTemplateTest{
+		chartPath: chartPath,
+		release:   "ccsm-helm-test",
+		namespace: "ccsm-helm-" + strings.ToLower(random.UniqueId()),
+		templates: []string{"charts/operate/templates/configmap.yaml"},
+	})
+}
+
+func (s *configmapTemplateTest) TestContainerGoldenTestConfigMapDefaults() {
+	options := &helm.Options{
+		KubectlOptions: k8s.NewKubectlOptions("", "", s.namespace),
+	}
+	output := helm.RenderTemplate(s.T(), options, s.chartPath, s.release, s.templates)
+	actual := []byte(output)
+
+	goldenFile := "configmap.golden.json"
+
+	if *update {
+		err := ioutil.WriteFile(goldenFile, actual, 0644)
+		s.Require().NoError(err, "Golden file was not writable")
+	}
+
+	expected, err := ioutil.ReadFile(goldenFile)
+
+	// then
+	s.Require().NoError(err, "Golden file doesn't exist or was not readable")
+	s.Require().Equal(string(expected), output)
```

Comment:
:question: What's the output when they aren't equal? I think the testify library comes with a `JSONEq` method, maybe this provides more specialized output/comparison? It also has a `YAMLEq` as well, by the way.

Reply:
Thanks good to know! I check it and actually I like the failure message better than the yamleq

String eq
```
        	Error Trace:	configmap_test.go:57
        	Error:      	Not equal: 
        	            	expected: "---\n# Source: ccsm-helm/charts/operate/templates/configmap.yaml\nkind: ConfigMap\nmetadata:\n  name: ccsm-helm-test-operate\n  labels:\n    app: camunda-cloud-self-managed\n    app.kubernetes.io/name: operate\n    helm.sh/chart: operate-0.0.14\n    app.kubernetes.io/instance: ccsm-helm-test\n    app.kubernetes.io/managed-by: Helm\n    app.kubernetes.io/version: 1.3.1\n    app.kubernetes.io/part-of: camunda-cloud-self-managed\n    app.kubernetes.io/component: operate\napiVersion: v1\ndata:\n  application.yml: |\n    # Operate configuration file\n    camunda.operate:\n      # ELS instance to store Operate data\n      elasticsearch:\n        # Cluster name\n        clusterName: elasticsearch\n        # Host\n        host: elasticsearch-master\n        # Transport port\n        port: 9200\n      # Zeebe instance\n      zeebe:\n        # Broker contact point\n        brokerContactPoint: \"ccsm-helm-test-zeebe-gateway:26500\"\n      # ELS instance to export Zeebe data to\n      zeebeElasticsearch:\n        # Cluster name\n        clusterName: elasticsearch\n        # Host\n        host: elasticsearch-master\n        # Transport port\n        port: 9200\n        # Index prefix, configured in Zeebe Elasticsearch exporter\n        prefix: zeebe-record\n    logging:\n      level:\n        ROOT: INFO\n        org.camunda.operate: DEBUG\n    #Spring Boot Actuator endpoints to be exposed\n    management.endpoints.web.exposure.include: health,info,conditions,configprops,prometheus"
        	            	actual  : "---\n# Source: ccsm-helm/charts/operate/templates/configmap.yaml\nkind: ConfigMap\nmetadata:\n  name: ccsm-helm-test-operate\n  labels:\n    app: camunda-cloud-self-managed\n    app.kubernetes.io/name: operate\n    helm.sh/chart: operate-0.0.14\n    app.kubernetes.io/instance: ccsm-helm-test\n    app.kubernetes.io/managed-by: Helm\n    app.kubernetes.io/version: 1.3.1\n    app.kubernetes.io/part-of: camunda-cloud-self-managed\n    app.kubernetes.io/component: operate\napiVersion: v1\ndata:\n  application2.yml: |\n    # Operate configuration file\n    camunda.operate:\n      # ELS instance to store Operate data\n      elasticsearch:\n        # Cluster name\n        clusterName: elasticsearch\n        # Host\n        host: elasticsearch-master\n        # Transport port\n        port: 9200\n      # Zeebe instance\n      zeebe:\n        # Broker contact point\n        brokerContactPoint: \"ccsm-helm-test-zeebe-gateway:26500\"\n      # ELS instance to export Zeebe data to\n      zeebeElasticsearch:\n        # Cluster name\n        clusterName: elasticsearch\n        # Host\n        host: elasticsearch-master\n        # Transport port\n        port: 9200\n        # Index prefix, configured in Zeebe Elasticsearch exporter\n        prefix: zeebe-record\n    logging:\n      level:\n        ROOT: INFO\n        org.camunda.operate: DEBUG\n    #Spring Boot Actuator endpoints to be exposed\n    management.endpoints.web.exposure.include: health,info,conditions,configprops,prometheus"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -16,3 +16,3 @@
        	            	 data:
        	            	-  application.yml: |
        	            	+  application2.yml: |
        	            	     # Operate configuration file
        	Test:       	TestConfigmapTemplate/TestContainerGoldenTestConfigMapDefaults
```

YamlEq:

```
        	Error Trace:	configmap_test.go:56
        	Error:      	Not equal: 
        	            	expected: map[string]interface {}{"apiVersion":"v1", "data":map[string]interface {}{"application.yml":"# Operate configuration file\ncamunda.operate:\n  # ELS instance to store Operate data\n  elasticsearch:\n    # Cluster name\n    clusterName: elasticsearch\n    # Host\n    host: elasticsearch-master\n    # Transport port\n    port: 9200\n  # Zeebe instance\n  zeebe:\n    # Broker contact point\n    brokerContactPoint: \"ccsm-helm-test-zeebe-gateway:26500\"\n  # ELS instance to export Zeebe data to\n  zeebeElasticsearch:\n    # Cluster name\n    clusterName: elasticsearch\n    # Host\n    host: elasticsearch-master\n    # Transport port\n    port: 9200\n    # Index prefix, configured in Zeebe Elasticsearch exporter\n    prefix: zeebe-record\nlogging:\n  level:\n    ROOT: INFO\n    org.camunda.operate: DEBUG\n#Spring Boot Actuator endpoints to be exposed\nmanagement.endpoints.web.exposure.include: health,info,conditions,configprops,prometheus"}, "kind":"ConfigMap", "metadata":map[string]interface {}{"labels":map[string]interface {}{"app":"camunda-cloud-self-managed", "app.kubernetes.io/component":"operate", "app.kubernetes.io/instance":"ccsm-helm-test", "app.kubernetes.io/managed-by":"Helm", "app.kubernetes.io/name":"operate", "app.kubernetes.io/part-of":"camunda-cloud-self-managed", "app.kubernetes.io/version":"1.3.1", "helm.sh/chart":"operate-0.0.14"}, "name":"ccsm-helm-test-operate"}}
        	            	actual  : map[string]interface {}{"apiVersion":"v1", "data":map[string]interface {}{"application2.yml":"# Operate configuration file\ncamunda.operate:\n  # ELS instance to store Operate data\n  elasticsearch:\n    # Cluster name\n    clusterName: elasticsearch\n    # Host\n    host: elasticsearch-master\n    # Transport port\n    port: 9200\n  # Zeebe instance\n  zeebe:\n    # Broker contact point\n    brokerContactPoint: \"ccsm-helm-test-zeebe-gateway:26500\"\n  # ELS instance to export Zeebe data to\n  zeebeElasticsearch:\n    # Cluster name\n    clusterName: elasticsearch\n    # Host\n    host: elasticsearch-master\n    # Transport port\n    port: 9200\n    # Index prefix, configured in Zeebe Elasticsearch exporter\n    prefix: zeebe-record\nlogging:\n  level:\n    ROOT: INFO\n    org.camunda.operate: DEBUG\n#Spring Boot Actuator endpoints to be exposed\nmanagement.endpoints.web.exposure.include: health,info,conditions,configprops,prometheus"}, "kind":"ConfigMap", "metadata":map[string]interface {}{"labels":map[string]interface {}{"app":"camunda-cloud-self-managed", "app.kubernetes.io/component":"operate", "app.kubernetes.io/instance":"ccsm-helm-test", "app.kubernetes.io/managed-by":"Helm", "app.kubernetes.io/name":"operate", "app.kubernetes.io/part-of":"camunda-cloud-self-managed", "app.kubernetes.io/version":"1.3.1", "helm.sh/chart":"operate-0.0.14"}, "name":"ccsm-helm-test-operate"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -3,3 +3,3 @@
        	            	  (string) (len=4) "data": (map[string]interface {}) (len=1) {
        	            	-  (string) (len=15) "application.yml": (string) (len=823) "# Operate configuration file\ncamunda.operate:\n  # ELS instance to store Operate data\n  elasticsearch:\n    # Cluster name\n    clusterName: elasticsearch\n    # Host\n    host: elasticsearch-master\n    # Transport port\n    port: 9200\n  # Zeebe instance\n  zeebe:\n    # Broker contact point\n    brokerContactPoint: \"ccsm-helm-test-zeebe-gateway:26500\"\n  # ELS instance to export Zeebe data to\n  zeebeElasticsearch:\n    # Cluster name\n    clusterName: elasticsearch\n    # Host\n    host: elasticsearch-master\n    # Transport port\n    port: 9200\n    # Index prefix, configured in Zeebe Elasticsearch exporter\n    prefix: zeebe-record\nlogging:\n  level:\n    ROOT: INFO\n    org.camunda.operate: DEBUG\n#Spring Boot Actuator endpoints to be exposed\nmanagement.endpoints.web.exposure.include: health,info,conditions,configprops,prometheus"
        	            	+  (string) (len=16) "application2.yml": (string) (len=823) "# Operate configuration file\ncamunda.operate:\n  # ELS instance to store Operate data\n  elasticsearch:\n    # Cluster name\n    clusterName: elasticsearch\n    # Host\n    host: elasticsearch-master\n    # Transport port\n    port: 9200\n  # Zeebe instance\n  zeebe:\n    # Broker contact point\n    brokerContactPoint: \"ccsm-helm-test-zeebe-gateway:26500\"\n  # ELS instance to export Zeebe data to\n  zeebeElasticsearch:\n    # Cluster name\n    clusterName: elasticsearch\n    # Host\n    host: elasticsearch-master\n    # Transport port\n    port: 9200\n    # Index prefix, configured in Zeebe Elasticsearch exporter\n    prefix: zeebe-record\nlogging:\n  level:\n    ROOT: INFO\n    org.camunda.operate: DEBUG\n#Spring Boot Actuator endpoints to be exposed\nmanagement.endpoints.web.exposure.include: health,info,conditions,configprops,prometheus"
        	            	  },
        	Test:       	TestConfigmapTemplate/TestContainerGoldenTestConfigMapDefaults
    --- FAIL: TestConfigmapTemplate/TestContainerGoldenTestConfigMapDefaults (0.17s)
```

It is even more problematic when something in application.yaml changed. I will keep string eq for now, lets see maybe we have to change it later.